Fixes #38962 - Update DateTimePicker to pf5

MariaAga opened the pull request:
The original task states:
Deprecate the 3 components: TimePicker, DatePicker and DateTimePicker
Investigate where the components are used so we can plan the work to replace their usage

But the components are used in erb forms, so they should still be wrapped. 
I only found them being used in the generate report erb. but in the future we might need to reuse the datetime picker, which PF doesnt supply directly.
foreman_remote_job_execution uses a different type of a datetime picker, thats because in there users can select a date and a time separately, but in some form, it might be necessary to have the date and time in the same text field, like in the generate page.

MariaAga committed:
Fixes #38962 - Update DateTimePicker to pf5

Refs #38962 - move clear,today buttons to be inside datetimepicker

MariaAga commented:
TODO: fix 
<img width="418" height="64" alt="image" src="https://github.com/user-attachments/assets/fa56c7bc-d8ee-42b5-b42c-7952e21140f9" />
 and add a validator to allow only future date (if prop is passed)

MariaAga commented:
Fixed review comments, added isFutureOnly prop, fixed the extra 0 bug

ekohl commented on webpack/assets/javascripts/react_app/components/common/DateTimePicker/DateTimePicker.js:
```diff
@@ -1,124 +1,188 @@
+  onChange, // additional prop to allow for custom onChange logic
+}) => {
+  const [isCalendarOpen, setIsCalendarOpen] = useState(false);
+  const [valueDate, setValueDate] = useState(formatDate(initialValue));
+  const [valueTime, setValueTime] = useState(formatTime(initialValue));
```
I don't mean to block the review here and certainly not block, but I notice you store a formatted date in the store. Isn't it possible to store a `Date` object instead of a string? My main concern is that you can't deal with timezones this way which can be important.

MariaAga replied:
Can you give an example? formatDate and formatTime use a `New Date(` object

ekohl replied:
When I read the flow you take `initialValue` (which could be timezone aware) deconstruct that into simple strings for both date and time. Then reuse that into a string `value`. Could the `TextInput` itself be fed a `Date` instance with a format function? Effectively removing the `formatDate` and `formatTime` helpers.

Out of my interest I tried to read up on what HTML can do. https://developer.mozilla.org/en-US/docs/Web/HTML/Reference/Elements/input/datetime-local uses the value `2018-06-12T19:30` which is the ISO 8601 format. So perhaps pass that around as much as possible?

Then I don't know what PatternFly does, but I noticed https://www.patternfly.org/components/forms/text-input/ supports `datetime-local` too (instead of `text`). Is that only in PF6?

And this is still for me to understand how we deal with dates, not a formal review.

MariaAga replied:
I will say that from my tests by editing the field with a different timezone all worked as expected, and the server respected the correct timezone (set the generate report to 3 mins in the future and it only downloaded 3 mins later)

pondrejk commented:
Heya, looked at packit, not sure how much in progress this is, but here are some observations anyway:
- there's one downstream test related, the way it is written now doesn't break robottelo, though not sure if it would make sense for users to display time and date in one field and then time again in a separate field
<img width="426" height="62" alt="image" src="https://github.com/user-attachments/assets/221851dc-0a29-432c-a3e5-974236b93bf2" />

- if we drop time from the first field I'd need to make some adjustments in airgun which is ok, but I'd need unique id for the time input field as well

- I cant select from the calendar popup and the popup doesn't disappear if I click away from it (have to click calendar icon again to roll it up)
<img width="464" height="431" alt="image" src="https://github.com/user-attachments/assets/a2141176-2fb7-4dc9-847a-cc265ee6863a" />

- clear all button clears the first field only
<img width="464" height="61" alt="image" src="https://github.com/user-attachments/assets/09fcbc01-5852-469b-9fc6-d0ca06fc062b" />

MariaAga commented:
@pondrejk 

> clear all button clears the first field only

Bug in PF https://github.com/patternfly/patternfly-react/issues/6569 
There is no way to empty the time field. I'll take a look if I can workaround it.

> I cant select from the calendar popup 

Fixed 

> and the popup doesn't disappear if I click away from it (have to click calendar icon again to roll it up)

I based my work on the PF design, and it doesnt disappear on click away as well
https://v5-archive.patternfly.org/components/date-and-time/date-and-time-picker

pondrejk commented:
> https://v5-archive.patternfly.org/components/date-and-time/date-and-time-picker

strangely that second example there with Date and time range picker doesn't have this problem, not sure why

MariaAga commented:
> strangely that second example there with Date and time range picker doesn't have this problem, not sure why

@pondrejk 
Is there a preference to close on clickout side? I can add it if thats what we expect

MariaAga commented:
todo: editing the time when date is empty should populate the date

pondrejk commented:
> Is there a preference to close on clickout side? I can add it if thats what we expect

I suspect users could see it as a bug, for example if they change focus without changing date and the popover will prevent interaction with another field/dropdown, something like here with time select:

<img width="477" height="152" alt="image" src="https://github.com/user-attachments/assets/ea94db56-d101-4929-b84a-2b0cb52ca93a" />

MariaAga commented:
Fixed: calendar close on clickout side and on date select, selecting a time when there is no date sets the date to today.

pondrejk commented on webpack/assets/javascripts/react_app/components/common/DateTimePicker/DateTimePicker.js:
```diff
@@ -1,124 +1,195 @@
+          </InputGroupItem>
+          <InputGroupItem className="date-picker-input-item">
+            <TimePicker
+              value={valueTime}
```
nitpick, maybe some ouiaId here as well?, because right now the time picker input has id=""

MariaAga replied:
there is no ouiaId prop for timepicker, instead I gave it 
```
id={`${id}-time-picker`}
```

pondrejk reviewed (approved):
ack from the functional side, UI works nice, the time offset is honored when creating a report. Also positive PRT result in https://github.com/SatelliteQE/robottelo/pull/20541 proves the automation is not affected and the tested scenario still works

ofedoren reviewed (dismissed):
Thanks, @MariaAga and @pondrejk !

Just for future reference sake:
 1. Why is the time shown twice? In patternfly docs there are examples where it's either shown together or completely separate. We currently have it mixed.
 2. The old UI had a `Today` button, but it doesn't seem to be present now. Does clear button do it instead?

MariaAga commented:
> Why is the time shown twice? In patternfly docs there are examples where it's either shown together or completely separate. We currently have it mixed.

I didnt like that the PF docs created a new time selector which was just a dropdown with times, and I wanted to reuse existing components, so I put the timepicker component as it is.
The time needs to be in the date-time field so the form will have a date-time value, the second time field is mostly for comfort and editing. 

> The old UI had a Today button, but it doesn't seem to be present now. Does clear button do it instead?

No, clear only removes the date, PF example just have buttons outside the date picker to reset and clear the date https://v5-archive.patternfly.org/components/date-and-time/date-picker#controlled-required 
Not sure if we want more buttons on the form, but what about something like this? (only for the date time picker)
<img width="593" height="535" alt="image" src="https://github.com/user-attachments/assets/b4d6c3e8-1aa1-493b-a356-daa05f7e198c" />

ofedoren commented:
> what about something like this? (only for the date time picker)

Honestly, I'm not sure if this button is being used by the users, but at least it would feel the same since we already have such button.

MariaAga commented:
Added a today button inside the datepicker. Moved the clear button for consistency, and also since the clear only clears the date 
<img width="593" height="535" alt="Screenshot From 2025-12-19 16-16-53" src="https://github.com/user-attachments/assets/c7578cd1-e9da-4f3c-aa7a-15222c60866e" />

pondrejk commented:
checked again with recent changes, rerun prt, I think this is good to go? @ofedoren ?

ofedoren reviewed (commented):
Thanks, @MariaAga, could you please squash the commits?

MariaAga commented:
squashed
